File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static ArrayList<Integer>[] graph;
    static boolean[] vis;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            graph[u].add(v);
            graph[v].add(u);
        }

        vis = new boolean[n + 1];
        int cnt = 0;

        for (int i = 1; i <= n; i++) {
            if (vis[i]) continue;
            bfs(i);
            cnt++;
        }

        System.out.print(cnt);
    }

    public static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        vis[start] = true;

        while (!queue.isEmpty()) {
            int cur = queue.poll();
            for (int next : graph[cur]) {
                if (!vis[next]) {
                    vis[next] = true;
                    queue.offer(next);
                }
            }
        }
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.kt
import java.util.*

fun main() {
    val vis = BooleanArray(1002)
    val (n, m) = readln().split(" ").map { it.toInt() }
    val graph = Array(n + 1) { mutableListOf<Int>() }
    repeat(m) {
        val num = readln().split(" ").map { it.toInt() }
        graph[num[0]].add(num[1])
        graph[num[1]].add(num[0])
    }
    var num = 0
    for (i in 1..n) {
        if (vis[i]) continue
        num++
        val q: Queue<Int> = LinkedList()
        q.add(i)
        vis[i] = true
        while (q.isNotEmpty()) {
            val cur = q.peek()
            q.poll()
            graph[cur].forEach {
                if (!vis[it]) {
                    q.add(it)
                    vis[it] = true
                }
            }
        }
    }
    println(num)
}